criu/mount.c

```diff
@@ -3821,3 +3821,133 @@ int remount_readonly_mounts(void)
 	 */
 	return call_helper_process(ns_remount_readonly_mounts, NULL);
 }
+
+enum tree_step {
+	STEP_STOP = -2,
+	STEP_CHILD = -1,
+	STEP_SIBLING = 0,
+	STEP_PARENT = 1,
+};
+
+static struct mount_info *mnt_subtree_step(struct mount_info *mi,
+					   struct mount_info *root,
+					   enum tree_step *step)
+{
+	*step = STEP_CHILD;
+
+	if (!list_empty(&mi->children))
+		return list_entry(mi->children.next, struct mount_info, siblings);
+
+	*step = STEP_SIBLING;
+
+	while (mi->parent && mi != root) {
+		if (mi->siblings.next == &mi->parent->children) {
+			mi = mi->parent;
+			*step += STEP_PARENT;
+		} else
+			return list_entry(mi->siblings.next, struct mount_info, siblings);
+	}
+
+	*step = STEP_STOP;
+	return NULL;
+}
+
+static void mnt_resort_siblings_full(struct mount_info *tree)
+{
+	struct mount_info *m, *p;
+	LIST_HEAD(list);
+
+	pr_info("\tResorting siblings full on %d\n", tree->mnt_id);
+	while (!list_empty(&tree->children)) {
+		m = list_first_entry(&tree->children, struct mount_info, siblings);
+		list_del(&m->siblings);
+
+		list_for_each_entry(p, &list, siblings)
+			if (strcmp(p->ns_mountpoint, m->ns_mountpoint) < 0)
+				break;
+
+		list_add_tail(&m->siblings, &p->siblings);
+		mnt_resort_siblings_full(m);
```

Comment:
recursion? what is a max depth of it? Do we really need it? Could you write a comment about what is going on here?

Reply:
Actually max depth is max depth of mount tree for a mntns. For the tree from images the depth is ok because we've already done mnt_resort_siblings on it with the same depth (Though I don't see any prove that recursion in mnt_resort_siblings will not fail). But for the restored tree (which is expected in case of some error in mount restore be a different tree) we can't say the depth is ok. So I think either 1) we need a depth check (likely non-recursive dfs) before calling mnt_resort_siblings* or 2) we need to do mnt_resort_siblings non-recursive way which will be harder because we at the same time need to alter the tree in these function (may be we can do it with our regular non-recursive dfs next, need to think more about it).

I prefer (1) but I'm not sure about safe limit on the depth of the recursion, not sure if just testing it will be enough. What do you think?

Reply:
Done (2) - removed recursion from both current patch and mnt_resort_siblings.

Reply:
Done (2) - removed recursion from both current patch and mnt_resort_siblings. But yet I have some problem with tests -> will try fixing it.